sandbox/src/index.ts

```diff
@@ -110,30 +190,88 @@ const adminUrl = process.env.ADMIN_URL as string;
 const repoUrl = process.env.REPO_URL as string;
 const javaBinPath = process.env.JAVA_PATH as string;
 
-async function loop() {
+const submissionApiUrl = urlJoin(adminUrl, 'api/submission');
+
+async function processNextSubmission(): Promise<SubmissionProcessingResult> {
 	let submissionData: SubmissionGetData | undefined;
 	try {
 		submissionData = await fetchQueuedSubmission();
-	} catch {
-		console.error('Failed to fetch submission');
-		return;
+	} catch (e) {
+		console.error(`Failed to fetch from ${submissionApiUrl} with error: ${e}`);
+		return SubmissionProcessingResult.Error;
 	}
 
 	if (!submissionData) {
 		console.error('Unable to fetch submission data');
-	} else {
-		try {
-			cloneAndRun(submissionData);
-		} catch {
-			console.error('Unable to clone and run');
-		}
+		return SubmissionProcessingResult.Error;
+	}
+
+	if (!submissionData.submission) {
+		return SubmissionProcessingResult.NoSubmissions;
+	}
+	
+	printSubmissionHeader(submissionData);
+
+	let processingResult: SubmissionProcessingResult;
+	try {
+		await cloneAndRun(submissionData);
+		processingResult = SubmissionProcessingResult.SubmissionProcessed;
+	} catch {
+		console.error('Failed to clone/run.');
+		processingResult = SubmissionProcessingResult.Error;
 	}
+
+	printSubmissionFooter(submissionData);
+	return processingResult;
+}
+
+function printSubmissionHeader(submissionData: SubmissionGetData) {
+	const submission = submissionData.submission;
+	if (!submission) {
+		return;
+	}
+
+	console.log(`--- Submission ${submission.id} ---`);
+	console.log(`- INFO: Contest ${submission.contestId} '${submission.contestName}', ` +
+		`Team ${submission.teamId} '${submission.teamName}', ` +
+		`Problem ${submission.problem.id} '${submission.problem.pascalName}', ` +
+		`SHA '${submission.commitHash}'`);
+}
+
+function printSubmissionFooter(submissionData: SubmissionGetData) {
+	const submission = submissionData.submission;
+	if (!submission) {
+		return;
+	}
+
+	console.log(`--- End Submission ${submission.id} ---`);
 }
 
 async function run() {
+	console.log("Sandbox started. Periodically checking for submissions.");
+
+	let iterationsSinceProcessedSubmission = 0;
+	let anySubmissionsProcessed = false;
 	while (true) {
-		await loop();
-		await new Promise((resolve) => setTimeout(resolve, 10000));
+		switch (await processNextSubmission()) {
+			case SubmissionProcessingResult.SubmissionProcessed:
+				iterationsSinceProcessedSubmission = 0;
+				anySubmissionsProcessed = true;
+				break;
+			case SubmissionProcessingResult.NoSubmissions:
+				if (iterationsSinceProcessedSubmission > 0 && iterationsSinceProcessedSubmission % 6 == 0) {
+					const numMinutes = iterationsSinceProcessedSubmission / 6;
+					console.log(`${numMinutes} minute${(numMinutes > 1 ? 's' : '')} since ` +
```

Comment:
Don't know how folks feel about regular log updates when nothing is happening just to see that it's still doing stuff. Maybe every 2 or 5 minutes would be better?

Reply:
Regular log updates sounds fine to me